src/D2L.CodeStyle.Analyzers/Immutability/ReadOnlyParameterAnalyzer.cs

```diff
@@ -32,77 +31,101 @@ internal sealed class ReadOnlyParameterAnalyzer : DiagnosticAnalyzer {
 				return;
 			}
 
-			context.RegisterSyntaxNodeAction(
-				ctx => AnalyzeParameter(
-					ctx: ctx,
-					readOnlyAttribute: readOnlyAttribute,
-					syntax: ctx.Node as ParameterSyntax
+			context.RegisterOperationAction(
+				ctx => AnalyzeMethodBodyOperation(
+					ctx,
+					readOnlyAttribute,
+					(IMethodBodyBaseOperation)ctx.Operation
+				),
+				OperationKind.ConstructorBody,
+				OperationKind.MethodBody
+			);
+
+			context.RegisterOperationAction(
+				ctx => AnalyzeLocalFunctionOperation(
+					ctx,
+					readOnlyAttribute,
+					(ILocalFunctionOperation)ctx.Operation
 				),
-				SyntaxKind.Parameter
+				OperationKind.LocalFunction
 			);
 		}
+		private static void AnalyzeMethodBodyOperation(
+			OperationAnalysisContext ctx,
+			INamedTypeSymbol readOnlyAttribute,
+			IMethodBodyBaseOperation operation
+		) => AnalyzeParameters(
+			ctx,
+			readOnlyAttribute,
+			(IMethodSymbol)ctx.ContainingSymbol,
+			operation.BlockBody ?? operation.ExpressionBody
+		);
 
-		private static void AnalyzeParameter(
-			SyntaxNodeAnalysisContext ctx,
+		private static void AnalyzeLocalFunctionOperation(
+			OperationAnalysisContext ctx,
 			INamedTypeSymbol readOnlyAttribute,
-			ParameterSyntax syntax
+			ILocalFunctionOperation operation
 		) {
-			SemanticModel model = ctx.SemanticModel;
-
-			if( syntax.AttributeLists.Count == 0 ) {
+			if( operation.Body is null ) {
 				return;
 			}
 
-			IParameterSymbol parameter = model.GetDeclaredSymbol( syntax, ctx.CancellationToken );
+			AnalyzeParameters(
+				ctx,
+				readOnlyAttribute,
+				operation.Symbol,
+				operation.Body
+			);
+		}
+
+		private static void AnalyzeParameters(
+			OperationAnalysisContext ctx,
+			INamedTypeSymbol readOnlyAttribute,
+			IMethodSymbol method,
+			IBlockOperation operation
+		) {
+			IParameterSymbol[] readOnlyParameters = method.Parameters.Where( p => IsMarkedReadOnly( readOnlyAttribute, p ) ).ToArray();
 
-			if( !IsMarkedReadOnly( readOnlyAttribute, parameter ) ) {
+			if( readOnlyParameters.Length == 0 ) {
 				return;
 			}
 
-			if( parameter.RefKind != RefKind.None ) {
-				/**
-				 * public void Foo( [ReadOnly] in int foo )
-				 * public void Foo( [ReadOnly] ref int foo )
-				 * public void Foo( [ReadOnly] out int foo )
-				 */
-				ctx.ReportDiagnostic(
-					Diagnostics.ReadOnlyParameterIsnt,
-					syntax.GetLocation(),
-					messageArgs: new[] { "is an in/ref/out parameter" }
-				);
-			}
-
-			IMethodSymbol method = parameter.ContainingSymbol as IMethodSymbol;
-			SyntaxNode methodSyntaxNode = method.DeclaringSyntaxReferences[ 0 ].GetSyntax( ctx.CancellationToken );

```

Comment:
`method.DeclaringSyntaxReferences` is always the declaration (`partial void Foo();`), never the implementation part (`partial void Foo() {}`), so getting the right syntax can be tricky. Plus we don't want duplicates, so would only want to raise the diagnostic (or do the work) on one ParameterSyntax or the other.

Acting on the operation gets us the right syntax and symbol, because we're starting from the _block_ rather than the declaration of a method.

This should be more performant because we're at the operation level rather than going from syntax to symbols, and we're doing the dataflow analysis once per method instead of once per readonly-parameter.